cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -562,6 +563,33 @@ Result<std::shared_ptr<Schema>> ExtractSchemaToBind(const compute::Declaration&
   } else if (declr.factory_name == "filter") {
     auto input_declr = std::get<compute::Declaration>(declr.inputs[0]);
     ARROW_ASSIGN_OR_RAISE(bind_schema, ExtractSchemaToBind(input_declr));
+  } else if (declr.factory_name == "project") {
+    auto input_declr = std::get<compute::Declaration>(declr.inputs[0]);
+    ARROW_ASSIGN_OR_RAISE(auto input_schema, ExtractSchemaToBind(input_declr));
+    const int num_fields_before_proj = input_schema->num_fields();
+    const auto& opts = checked_cast<const compute::ProjectNodeOptions&>(*(declr.options));
+    const auto& exprs = opts.expressions;
+    int i = 0;
+    bind_schema = input_schema;
+    for (const auto& expr : exprs) {
+      std::shared_ptr<Field> project_field;
+      auto bound_expr = expr.Bind(*input_schema);
+      if (auto* expr_call = bound_expr->call()) {
+        project_field = field(expr_call->function_name,
+                              expr_call->kernel->signature->out_type().type());
```

Comment:
Is there a `expr_call->type`?

Reply:
You're right. That's way straightforward and neater. We can use the `TypeHolder` and call `GetSharedPtr`.